### ISSUE TITLE
PERF: detect LayerNorm pattern and lower to mx.fast.layer_norm (2.4× speedup)

### ISSUE THREAD

**Issue by cetagostini** [pymc-devs/pytensor#2088]
Part of #2085.

## Current implementation

PyTensor builds LayerNorm out of basic ops: `(x - mean(x)) * rsqrt(var(x) + eps) * gamma + beta`. The MLX backend dispatches each elementwise/reduction op separately and lets `mx.compile` fuse them. `mx.compile` does fuse this into a small number of Metal kernels, but **not** as efficiently as MLX's hand-written `mx.fast.layer_norm`, which is a single kernel doing the same computation.

## Reproducer + metrics

```python
import time, statistics, mlx.core as mx, numpy as np

def sync(): mx.synchronize()

def bench(fn, *args, n=30, runs=5):
    sync()
    for _ in range(8): mx.eval(fn(*args))
    sync()
    out = []
    for _ in range(runs):
        sync(); t0 = time.perf_counter()
        for _ in range(n): mx.eval(fn(*args))
        sync()
        out.append((time.perf_counter() - t0) / n)
    return statistics.median(out) * 1e6

B, S, D = 8, 512, 2048
rng = np.random.default_rng(0)
x = mx.array(rng.standard_normal((B, S, D)).astype(np.float32))
g = mx.array(np.ones(D, dtype=np.float32))
b = mx.array(np.zeros(D, dtype=np.float32))
mx.eval(x, g, b); sync()

@mx.compile
def manual(x, g, b):
    m = x.mean(axis=-1, keepdims=True)
    v = x.var(axis=-1, keepdims=True)
    return (x - m) * mx.rsqrt(v + 1e-5) * g + b

@mx.compile
def fast(x, g, b):
    return mx.fast.layer_norm(x, weight=g, bias=b, eps=1e-5)

print(f"manual: {bench(manual, x, g, b):.1f} us")
print(f"fast:   {bench(fast,   x, g, b):.1f} us")
```

| Implementation | Median (us) |
|---|---|
| Manual (composed ops, what PyTensor produces) | **1067** |
| `mx.fast.layer_norm` | **442** |

→ **2.4× speedup**. Numerics match within ~3e-6 (fp32 vs fused).

## Proposed change

Two pieces:

1. **A new MLX-only Op** `MLXLayerNorm(eps)` in `pytensor/link/mlx/ops.py` (new file). Funcify dispatches to `mx.fast.layer_norm`.

2. **A node rewriter** registered with the MLX-specific `mlx` rewrite tag that detects the pattern and emits `MLXLayerNorm`. Same pattern PyTensor already uses for backend-specialised rewrites (cf. JAX-only rewrites referenced in `JAXLinker.required_rewrites`).

**Files:**
- `pytensor/link/mlx/ops.py` (new) — defines `MLXLayerNorm` Op
- `pytensor/link/mlx/rewriting/__init__.py` (new) — registers `mlx` rewrite tag
- `pytensor/link/mlx/rewriting/fused_norm.py` (new) — pattern matcher
- `pytensor/link/mlx/dispatch/__init__.py` — import the new dispatch
- `pytensor/compile/mode.py` — register `mlx` rewrites in `MLX` mode

### Sketch

```python
# pytensor/link/mlx/ops.py
from pytensor.graph.op import Op

class MLXLayerNorm(Op):
    __props__ = ("eps",)
    def __init__(self, eps): self.eps = eps
    def make_node(self, x, weight, bias):
        out = x.type()
        return Apply(self, [x, weight, bias], [out])
    def perform(self, node, inputs, outputs):
        # Required for fallback, real work happens in mlx_funcify
        x, w, b = inputs
        m = x.mean(axis=-1, keepdims=True)
        v = x.var(axis=-1, keepdims=True)
        outputs[0][0] = (x - m) / np.sqrt(v + self.eps) * w + b


# pytensor/link/mlx/dispatch/fused.py
@mlx_funcify.register(MLXLayerNorm)
def mlx_funcify_MLXLayerNorm(op, **kwargs):
    eps = op.eps
    def layer_norm(x, weight, bias):
        return mx.fast.layer_norm(x, weight=weight, bias=bias, eps=eps)
    return layer_norm


# pytensor/link/mlx/rewriting/fused_norm.py
from pytensor.graph.rewriting.basic import node_rewriter
@node_rewriter(tracks=[Elemwise])
def fuse_layer_norm(fgraph, node):
    """Detect (x - mean(x, -1, kd)) / sqrt(var(x, -1, kd) + eps) * g + b → MLXLayerNorm."""
    matched = match_layer_norm_subgraph(node)  # standard pattern walker
    if matched is None:
        return None
    x, gamma, beta, eps = matched
    return [MLXLayerNorm(eps=eps)(x, gamma, beta)]
```

Pattern matching can reuse approaches from existing rewrites such as the softmax-with-crossentropy detector.

### Gradient

`mx.fast.layer_norm` supports backward via `mx.value_and_grad`, so registering a `grad` method on `MLXLayerNorm` that returns the same composite gradient (or another fused op) keeps PyTensor's `pt.grad` working transparently. Even with the Pythonic gradient, the forward fast kernel is the win.

## Acceptance criteria

- [ ] `MLXLayerNorm` Op defined and dispatched.
- [ ] Pattern rewrite fires only when `mode="MLX"` (via `mlx` tag).
- [ ] Numerics test: `pt.function(...)` graph that builds the LayerNorm pattern produces results matching `mx.fast.layer_norm` within `1e-5`.
- [ ] Benchmark in `tests/link/mlx/` shows ≥ 2× speedup over the unpatterned baseline at the shape above.

**Comment by jessegrabowski** [comment]
I'm closing these NN specific suggestions because something like this is out of scope for pytensor. We removed the `.nn` module to focus on providing a more general computation engine (scipy/numpy) rather than a deep learning framework.

PRs welcome to https://github.com/jessegrabowski/pytensor_ml, or if you want to spin up your own I'd be happy to contribute.
